### PR TITLE
Docs for signing off-chain messages

### PR DESCRIPTION
**Pull request by drknzz**
### **Please check if the PR fulfills these requirements**
- [x] The formatter, linter, and tests all run without an error
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes / features)

## **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Add section in docs for signing off-chain messages
Fix creating typed data dictionary through TypedData TypedDict

### CODE REVIEW

File: starknet_py/tests/e2e/docs/guide/test_sign_offchain_message.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    # Create an AccountClient instance
+    client = GatewayClient("testnet")
+    account_client = AccountClient(
+        client=client,
+        address="0x7536539dbba2a49ab688a1c86332625f05f660a94908f362d29212e6071432d",
+        chain=StarknetChainId.TESTNET,
+        key_pair=KeyPair.from_private_key(
+            2640601739596882773716581229189050365310622467205517187262055898295170952602
+        ),
```
Comment (war-in) [inline starknet_py/tests/e2e/docs/guide/test_sign_offchain_message.py:56]: I would not use a real testnet address and private key. This should be set to some dummy value ("0x2137") and then changed to the parametrized one

**Review by war-in — CHANGES_REQUESTED** [review]

File: starknet_py/tests/e2e/docs/guide/test_sign_offchain_message.py (hunk 1 above)
Comment (cptartur) [reply starknet_py/tests/e2e/docs/guide/test_sign_offchain_message.py:56]: Agreed, this test should be runnable on devnet (and other nets) like all other docs tests.

File: starknet_py/tests/e2e/docs/guide/test_sign_offchain_message.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        "contents": "Hello, Bob!",
+    }
+
+    # Create a TypedData dictionary
+    typed_data = TypedData(
+        types=types,
+        primaryType="Mail",
+        domain=starknet_domain,
+        message=message,
+    )
```
Comment (cptartur) [inline starknet_py/tests/e2e/docs/guide/test_sign_offchain_message.py:46]: Could we replace these steps with the creation of a single standard dict to show it is possible to use `sign_message` without explicitly instantiating TypedData typeddict?

Something like
```python
typed_data = {
  "types": {...},
  "primary_type": "Mail"
  "domain": {...},
  "message: {...}
}
```

Current approach suggests it's necessary to create all these objects.

**Review by cptartur — CHANGES_REQUESTED** [review]

**Review by war-in — APPROVED** [review]

File: starknet_py/net/account/account_client.py [hunk 3]
```diff
@@ -629,7 +629,7 @@ async def _verify_message_hash(self, msg_hash: int, signature: List[int]) -> boo
             nonce=None,
         )
         try:
-            await self.call_contract(invoke_tx=invoke_tx)
+            await self.call_contract(invoke_tx=invoke_tx, block_hash="latest")
```
Comment (cptartur) [inline starknet_py/net/account/account_client.py:632]: Why was this added?

File: starknet_py/tests/e2e/conftest.py [hunk 4]
```diff
@@ -52,6 +52,9 @@
     "0X126FAB6AE8ACA83E2DD00B92F94F3402397D527798E18DC28D76B7740638D23"
 )
 
+directory_with_contracts = Path(os.path.dirname(__file__)) / "mock_contracts_dir"
+mocks_dir = Path(os.path.dirname(__file__)) / "mocks"
```
Comment (cptartur) [inline starknet_py/tests/e2e/conftest.py:56]: If you want to name your directory `mocks` perhaps we should consider renaming `mock_contracts_dir` to just `contracts`?

Comment (cptartur) [reply starknet_py/tests/e2e/conftest.py:56]: cc @war-in

File: starknet_py/tests/e2e/docs/guide/test_sign_offchain_message.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    # We can calculate the message hash
+    msg_hash = account_client.hash_message(typed_data=typed_data)
+
+    # Sign the message
+    signature = account_client.sign_message(typed_data=typed_data)
+
+    # Verify the message
+    assert await account_client.verify_message(
+        typed_data=typed_data, signature=signature
```
Comment (cptartur) [inline starknet_py/tests/e2e/docs/guide/test_sign_offchain_message.py:75]: I think we shouldn't be using `assert` in these examples in case someone uses this code in production, as asserts can be disabled there.

cc @war-in

Comment (war-in) [reply starknet_py/tests/e2e/docs/guide/test_sign_offchain_message.py:75]: You're right! I would move assert to the end of the file after # add to docs: end

File: starknet_py/tests/e2e/conftest.py (hunk 4 above)
Comment (cptartur) [reply starknet_py/tests/e2e/conftest.py:56]: I just think having `mocks/` and `mock_contracts_dir/` is kind of confusing. For me it's fine to have "mock" contracts in `contracts/`, but if you don't like it, we don't have to go with that approach.

File: starknet_py/net/account/account_client.py (hunk 3 above)
Comment (drknzz) [reply starknet_py/net/account/account_client.py:632]: Without it, full node client had a problem in rpc_call as devnet does not support other blocks than latest.

**Review by cptartur — APPROVED** [review]
